### ISSUE TITLE
refactor for avoiding unnecessary checks when _reduce parameter is not a transformer.

### ISSUE THREAD

**Issue by xgbuils** [ramda/ramda#3247]
I was reading the things commented in the PR #2762 and I also agree with the idea that `_reduce` is too complex and maybe _reduced behavior should be relegated to the transducers and no more. However I think that the solution is adding more complexity and more checks at running time. It's true that we are gaining performance with no transducer way of work. But we are paying a cost of a new check when we work with tranducers.

My proposal is having a `_reduce` function that works with reducers and a `_xReduce` function that works with transformers. Maybe we will have a kind of code duplication. But I think it's better this kind of duplication over adding checks at runtime. Then, I'm gonna use `_reduce` or `_xReduce` in the ramda functions depending on the case. My idea is breaking nothing. It's just keeping things simple and having a good enabler if we really want to avoid to manage reduced values in some of _dispatchable fallbacks.

**Comment by CrossEye** [comment]
I can't recall the history now (or bother trying to dig it up) so I don't know if `reduced` predated transducers or was added for that case, but it does add some nasty complexity.

Of course if I were to design it today, I would, ignoring FantasyLand compatibility, write something more like this <sup>1</sup>:

```js
const None = Symbol ()

const foldl = (fn) => (init) => ([x = None, ...xs]) =>
  x == None
    ? init
    : fn (init, x, (r) => foldl (fn) (r) (xs), (r) => r)


const total = foldl ((a, b, next, done) => next (a + b)) (0) 

total ([1, 2, 3, 4, 5, 6, 7]) //=> 28

const sumBelow = (t) => 
  foldl ((a, b, next, done) => a + b < t ? next (a + b) : done (a)) (0)

sumBelow (20) ([1, 2, 3, 4, 5, 6, 7]) //=> 15
```

But I'm afraid that ship sailed long ago.

So this sounds like a good idea.  But we do need to keep in mind that removing `reduced` from the plain reducers is a breaking change.


----------
<sup>1</sup> Or perhaps like this:
> ```js
> const foldl = (fn) => (init) => (xs) =>
>   xs .length < 1
>     ? init
>     : fn (init, xs [0] , (r) => foldl (fn) (r) (xs .slice (1)), (r) => r)
> ```

**Comment by xgbuils** [comment]
Hi @CrossEye 

I'm a little bit confused. Maybe I didn't explained well myself. I'm not trying to change the behaviour or reopen the #2762 discussion. Inspired by #2762 solution, I'm just trying to gain in performance removing extra checks in the code instead of adding them. I've sent the PR if you need a more specific examples of what I did.

> So this sounds like a good idea. But we do need to keep in mind that removing reduced from the plain reducers is a breaking change.

I can show a counterexample. At least, it's not always a breaking change. If we have the `liftN` implementation code in ramda:
``` javascript
var liftN = _curry2(function liftN(arity, fn) {
  var lifted = curryN(arity, fn);
  return curryN(arity, function() {
    return _reduce(ap, map(lifted, arguments[0]), Array.prototype.slice.call(arguments, 1));
  });
});
```
I can replace it to:
``` javascript
var liftN = _curry2(function liftN(arity, fn) {
  var lifted = curryN(arity, fn);
  return curryN(arity, function() {
    var args = Array.prototype.slice.call(arguments, 1);
    return args.reduce(ap, map(lifted, arguments[0]));
  });
});
```
I can replace the `_reduce` function by `Array.prototype.reduce` without any breaking change because `R.ap`  doesn't return reduced values and `Array.prototype.slice.call(arguments, 1)` is always an array.

I didn't exactly do that in the PR but it's an example of the idea of what I did. `R.reduce` and other functions that depends on reduced values behaves as before.

**Comment by CrossEye** [comment]
I think I just got stuck by the reference to #2762.  Thank you for the explanation.  This does sound like a good idea, and I'll look at the PR.

**Comment by xgbuils** [comment]
Closed by #3248
